### PR TITLE
Split files when planning scan tasks

### PR DESCRIPTION
**Pull request by rdsr**
Opening a new PR since I accidentally deleted the branch associated with the PR #111 
This addresses all the comments from #111 

New tests are added for Avro and Parquet.  There's a lot common among these tests. I was first thinking that I could write these tests under `data` module. I've decided not to do that yet, since these tests are not specific to `data`, but if you feel there's a good way to remove the redundancy I will be happy to address that.

The question of broken tests under `TestFilteredScan` is still open. I think @rdblue will look into them and  provide feedback on how to address it.

### CODE REVIEW

**Comment by rdsr** [comment]
cc: @rdblue

File: api/src/main/java/com/netflix/iceberg/FileScanTask.java [hunk 1]
```diff
@@ -64,6 +64,9 @@
    */
   Expression residual();
 
+
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/FileScanTask.java]: Nit: extra blank line.

File: core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java [hunk 2]
```diff
@@ -38,6 +41,7 @@
     this.residuals = residuals;
   }
 
+
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java]: Nit: non-functional change

File: core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java [hunk 3]
```diff
@@ -20,9 +20,12 @@
 package com.netflix.iceberg;
 
 import com.google.common.base.Objects;
+import com.google.common.collect.Lists;
 import com.netflix.iceberg.expressions.Expression;
 import com.netflix.iceberg.expressions.ResidualEvaluator;
 
+import java.util.Iterator;
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java]: Nit: please don't separate imports into groups.

File: core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java [hunk 4]
```diff
@@ -66,6 +70,15 @@ public Expression residual() {
     return residuals.residualFor(file.partition());
   }
 
+  @Override
+  public Iterable<FileScanTask> split(long splitSize) {
+    if (file.format().isSplittable()) {
+      return () -> new SplitScanTaskIterator(splitSize, this);
+    } else {
+      return Lists.newArrayList(this);
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java]: I think it would be better to use `ImmutableList.of(this)`. It is good to return immutable objects even though this probably won't be mutated. And that `ImmutableList` can use an implementation that is more efficient than `ArrayList` because it knows that there is only going to be one item.

File: spark/src/test/java/com/netflix/iceberg/spark/source/TestFilteredScan.java [hunk 5]
```diff
@@ -429,7 +429,9 @@ private File buildPartitionedTable(String desc, PartitionSpec spec, String udf,
     Table byId = TABLES.create(SCHEMA, spec, location.toString());
 
     // do not combine splits because the tests expect a split per partition
-    byId.updateProperties().set("read.split.target-size", "1").commit();
+    //TODO: this is commented out since the current patch will create
+    // too many small scan tasks
+    // byId.updateProperties().set("read.split.target-size", "1").commit();
```
Comment (rdblue) [inline spark/src/test/java/com/netflix/iceberg/spark/source/TestFilteredScan.java]: This was set to avoid combining splits. Instead of setting it to 1, it works to set it to some value larger than the file size to avoid splitting. But the value can't be too large, or else splits get combined. Looks like 2048 works so that no files are split or combined.

File: parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+
+import static com.netflix.iceberg.types.Types.NestedField.required;
+import static org.apache.avro.generic.GenericData.Record;
+
+public class TestParquetSplitScan {
```
Comment (rdblue) [inline parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java]: I think I'd rather add these tests in iceberg-data to avoid the duplication. It would also be better to use Iceberg's generics instead of Avro generics because I think we will want to remove the dependence on Avro's generics. (Possibly remove the dependency on Avro.)

File: parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        .build();
+
+    table.newAppend().appendFile(dataFile).commit();
+  }
+
+  private int addRecordsToFile(File file) throws IOException {
+    // With these number of records and the given schema
+    // we can effectively write a file of size ~ 64 MB
+    int nRecords = 1600000;
+    try (FileAppender<Record> writer = Parquet.write(Files.localOutput(file))
```
Comment (rdblue) [inline parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java]: There's a `RandomDataGenerator` in iceberg-data that would cut down on code here and provide a slightly better test.

File: core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java [hunk 8]
```diff
@@ -74,4 +87,74 @@ public String toString() {
         .add("residual", residual())
         .toString();
   }
+
+  private static final class SplitScanTaskIterator implements Iterator<FileScanTask> {
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java]: I'd like to see a unit test for this class that has a few example split cases, like file length = 100 and split size = 15 produces [0, 15), [15, 30), [30, 45), etc. That ensures there are no gaps between splits. Just reading data files doesn't make that guarantee.

File: core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java (hunk 4 above)
Comment (rdsr) [reply core/src/main/java/com/netflix/iceberg/BaseFileScanTask.java]: Makes sense

File: parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java (hunk 6 above)
Comment (rdsr) [reply parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java]: Thanks for the input. I can move these tests under `iceberg-data` module and remove duplication.

File: spark/src/test/java/com/netflix/iceberg/spark/source/TestFilteredScan.java (hunk 5 above)
Comment (rdsr) [reply spark/src/test/java/com/netflix/iceberg/spark/source/TestFilteredScan.java]: Thankyou :)

File: parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java (hunk 7 above)
Comment (rdsr) [reply parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java]: I first had the tests written with {{RandomDataGenerator}} and then I moved to  this thinking that there could be slight changes in data sizes which could result in the tests failing randomly. I don't think that should be the case though if I author the tests carefully though.

Comment (rdsr) [reply parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java]: Oh, another reason for not using `RandomDataGenerator` was that I can easily make sure that all records are read and are in order using the `id` field in the record, which is a monotonically increasing number.  
@rdblue I'd like to keep this approach for now, unless u feel otherwise

Comment (rdblue) [reply parquet/src/test/java/com/netflix/iceberg/parquet/TestParquetSplitScan.java]: `RandomDataGenerator` is passed a seed that is constant across runs to avoid random failures. You get a consistent test, but we can share the random data generation.

As for validating the order, it should be no different to use the generated data. It returns you a list of records that you write to a file. Then you loop through the list and verify that the record from the file is the same as the corresponding record from the list.

I think this should use the generator.

**Comment by rdsr** [comment]
Addressed pending comments.

File: core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+import static com.netflix.iceberg.BaseFileScanTask.SplitScanTaskIterator;
+
+public class TestSplitScanTaskIterator {
+  @Test
+  public void testSplits() {
+    long splitSize = 15L;
+    List<FileScanTask> tasks = Lists.newArrayList(new SplitScanTaskIterator(splitSize, new MockFileScanTask(100L)));
+    Assert.assertEquals(7, tasks.size());
+    long offset = 0;
+    for (int i = 0; i < tasks.size(); i++) {
```
Comment (rdblue) [inline core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java]: This should use a hard-coded set of expected splits to validate the results.

That is much easier to read and we don't have to worry about the correctness of the arithmetic in the tests as well as in the split code.

File: core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    Assert.assertEquals(tasks.get(0).start(), 0);
+    Assert.assertEquals(tasks.get(0).length(), 10);
+
+    tasks = Lists.newArrayList(new SplitScanTaskIterator(20, new MockFileScanTask(10)));
+    Assert.assertEquals(1, tasks.size());
+    Assert.assertEquals(tasks.get(0).start(), 0);
+    Assert.assertEquals(tasks.get(0).length(), 10);
+  }
+
+  private static class MockFileScanTask implements FileScanTask {
```
Comment (rdblue) [inline core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java]: Could this use `BaseFileScanTask` instead?

File: data/src/test/java/com/netflix/iceberg/TestSplitScan.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  @Test
+  public void test() {
+    Assert.assertEquals(
+        "There should be 4 tasks created since file size is approximately close to 64MB and split size 16MB",
+        4, Lists.newArrayList(table.newScan().planTasks()).size());
+    List<Record> records = Lists.newArrayList(IcebergGenerics.read(table).build());
+    Assert.assertEquals(expectedRecords.size(), records.size());
+    for (int i = 0; i < expectedRecords.size(); i++)
```
Comment (rdblue) [inline data/src/test/java/com/netflix/iceberg/TestSplitScan.java]: Nit: control statements like `if` and `for` should always use `{` and `}`.

File: spark/src/test/java/com/netflix/iceberg/spark/source/TestFilteredScan.java [hunk 12]
```diff
@@ -429,7 +429,7 @@ private File buildPartitionedTable(String desc, PartitionSpec spec, String udf,
     Table byId = TABLES.create(SCHEMA, spec, location.toString());
 
     // do not combine splits because the tests expect a split per partition
-    byId.updateProperties().set("read.split.target-size", "1").commit();
+    byId.updateProperties().set("read.split.target-size", "2048").commit();
```
Comment (rdblue) [inline spark/src/test/java/com/netflix/iceberg/spark/source/TestFilteredScan.java]: Can you update the comment to note that a lower value will also cause failures because the files are split? Something like "use a target size that will not combine or split files ..."

**Comment by rdblue** [comment]
@rdsr, looking great! Just one more issue with the split unit test and I think this is about ready. Thanks!

File: core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java (hunk 9 above)
Comment (rdsr) [reply core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java]: That makes a lot of sense @rdblue

File: core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java (hunk 10 above)
Comment (rdsr) [reply core/src/test/java/com/netflix/iceberg/TestSplitScanTaskIterator.java]: Possibly, I can extend `BaseFileScanTask` and use that..

File: data/src/test/java/com/netflix/iceberg/TestSplitScan.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        .withRecordCount(expectedRecords.size())
+        .withFileSizeInBytes(file.length())
+        .withPath(file.toString())
+        .withFormat(format)
+        .build();
+
+    table.newAppend().appendFile(dataFile).commit();
+  }
+
+  private List<Record> writeFile(File file, FileFormat format) throws IOException {
```
Comment (rdblue) [inline data/src/test/java/com/netflix/iceberg/TestSplitScan.java]: Minor: it's okay because this is a test, but methods should generally do one thing: either write records to a file or generate records, not both.

File: core/src/main/java/com/netflix/iceberg/BaseTableScan.java [hunk 14]
```diff
@@ -41,6 +42,7 @@
 import com.netflix.iceberg.util.ParallelIterable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseTableScan.java]: Please remove unnecessary whitespace additions.

Here's an example problem caused by this: https://github.com/apache/spark/commit/6d22ee3969597ed5e38333d5b6f0f891a859594e#r32770189

File: api/src/main/java/com/netflix/iceberg/FileScanTask.java [hunk 15]
```diff
@@ -64,6 +64,8 @@
    */
   Expression residual();
 
+  Iterable<FileScanTask> split(long splitSize);
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/FileScanTask.java]: Sorry I caught it late, but this needs Javadoc.

**Comment by rdblue** [comment]
@rdsr, this is really close to being ready! Unfortunately, I missed that there was no javadoc for `split` in the last review. So there's just that and the whitespace addition to fix before thsi goes in. Thanks!

File: core/src/main/java/com/netflix/iceberg/BaseTableScan.java (hunk 14 above)
Comment (rdsr) [reply core/src/main/java/com/netflix/iceberg/BaseTableScan.java]: My bad. I'll be more careful next time.

Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/BaseTableScan.java]: Not a problem, I just wanted to highlight that these do actually cause problems later on so it's best to catch them during reviews.

File: data/src/test/java/com/netflix/iceberg/TestSplitScan.java (hunk 13 above)
Comment (rdsr) [reply data/src/test/java/com/netflix/iceberg/TestSplitScan.java]: Fixed

**Comment by rdblue** [comment]
Merged. Thanks for fixing this, @rdsr!
